pkg/skaffold/render/renderer/renderer.go

```diff
@@ -19,56 +19,83 @@ import (
 	DefaultHydrationDir = ".kpt-pipeline"
 	dryFileName         = "manifests.yaml"
-	KptFileName         = "Kptfile"
 )
 
 type Renderer interface {
 	Render(context.Context, io.Writer, []graph.Artifact) error
 }
 
-func NewSkaffoldRenderer(config *latestV2.RenderConfig, workingDir string) Renderer {
+// NewSkaffoldRenderer creates a new Renderer object from the latestV2 API schema.
+func NewSkaffoldRenderer(config *latestV2.RenderConfig, workingDir string) (Renderer, error) {
 	// TODO(yuwenma): return instance of kpt-managed mode or skaffold-managed mode defer to the config.Path fields.
 	// The alpha implementation only has skaffold-managed mode.
-	generator := generate.NewGenerator(workingDir, *config.Generate)
-	return &SkaffoldRenderer{Generator: *generator, workingDir: workingDir}
+	// TODO(yuwenma): The current work directory may not be accurate if users use --filepath flag.
+	hydrationDir := filepath.Join(workingDir, DefaultHydrationDir)
+
+	var generator *generate.Generator
+	if config.Generate == nil {
+		// If render.generate is not given, default to current working directory.
+		defaultManifests := filepath.Join(workingDir, "*.yaml")
+		generator = generate.NewGenerator(workingDir, latestV2.Generate{Manifests: []string{defaultManifests}})
+	} else {
+		generator = generate.NewGenerator(workingDir, *config.Generate)
+	}
+
+	var validator *validate.Validator
+	if config.Validate != nil {
+		var err error
+		validator, err = validate.NewValidator(*config.Validate)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		validator, _ = validate.NewValidator([]latestV2.Validator{})
+	}
+	return &SkaffoldRenderer{Generator: *generator, Validator: *validator, workingDir: workingDir, hydrationDir: hydrationDir}, nil
 }
 
 type SkaffoldRenderer struct {
 	generate.Generator
-	workingDir string
-	labels     map[string]string
+	validate.Validator
+	workingDir   string
+	hydrationDir string
+	labels       map[string]string
 }
 
 // prepareHydrationDir guarantees the existence of a kpt-initialized temporary directory.
 // This directory is used to cache DRY config and hydrates the DRY config to WET config in-place.
 // This is needed because kpt v1 only supports in-place config while users may not want to have their config be
 // hydrated in place.
 func (r *SkaffoldRenderer) prepareHydrationDir(ctx context.Context) error {
-	// TODO(yuwenma): The current work directory may not be accurate if users use --filepath flag.
-	outputPath := filepath.Join(r.workingDir, DefaultHydrationDir)
-	logrus.Debugf("creating render directory: %v", outputPath)
-
-	if err := os.MkdirAll(outputPath, os.ModePerm); err != nil {
-		return fmt.Errorf("creating cache directory for hydration: %w", err)
+	if _, err := os.Stat(r.hydrationDir); os.IsNotExist(err) {
+		logrus.Debugf("creating render directory: %v", r.hydrationDir)
+		if err := os.MkdirAll(r.hydrationDir, os.ModePerm); err != nil {
+			return fmt.Errorf("creating cache directory for hydration: %w", err)
```

Comment:
```suggestion
			return fmt.Errorf("creating render directory for hydration: %w", err)
```

Reply:
Is this a fatal error? if its just a cache dir, we can still continue without caching?

Reply:
since kpt only supports in-place hydration, we do need a tmp dir to hold the DRY configs so as the hydrated config won't pollute the user working dir.